core/src/apps/cardano/address.py

```diff
@@ -1,145 +1,344 @@
-            raise NETWORK_MISMATCH
-    else:  # testnet
-        if len(attributes) == 0 or PROTOCOL_MAGIC_KEY not in attributes:
-            raise NETWORK_MISMATCH
 
-        protocol_magic_cbor = attributes[PROTOCOL_MAGIC_KEY]
-        address_protocol_magic = cbor.decode(protocol_magic_cbor)
+def _validate_address_network_id(address: bytes, network_id: int) -> None:
+    if _get_address_network_id(address) != network_id:
+        raise NETWORK_MISMATCH
+
+
+def _get_address_network_id(address: bytes) -> int:
+    return address[0] & 0x0F
+
+
+def get_public_key_hash(keychain: seed.Keychain, path: List[int]) -> bytes:
+    node = keychain.derive(path)
+    public_key = remove_ed25519_prefix(node.public_key())
+    return hashlib.blake2b(data=public_key, outlen=28).digest()
+
+
+def get_human_readable_address(address: bytes, network_id: int) -> str:
+    if _has_byron_address_header(address):
+        return base58.encode(address)
+    elif _has_shelley_address_header(address):
+        hrp = _get_bech32_hrp_for_address(_get_address_type(address), network_id)
+        return bech32.encode(hrp, address)
+    else:
+        raise ValueError("Invalid address")
+
+
+def derive_human_readable_address(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> str:
+    address = derive_address_bytes(keychain, parameters, protocol_magic, network_id)
+    return get_human_readable_address(address, network_id)
+
+
+def derive_address_bytes(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> bytes:
+    is_byron_address = parameters.address_type == CardanoAddressType.BYRON
+
+    if is_byron_address:
+        address = _derive_byron_address(
+            keychain, parameters.spending_key_path, protocol_magic
+        )
+    else:
+        address = _derive_shelley_address(keychain, parameters, network_id)
+
+    return address
 
-        if not isinstance(address_protocol_magic, int):
-            raise INVALID_ADDRESS
 
-        if address_protocol_magic != protocol_magic:
-            raise NETWORK_MISMATCH
+def _derive_byron_address(
+    keychain: seed.Keychain, path: List[int], protocol_magic: int
+) -> bytes:
+    if not is_byron_path(path):
+        raise wire.DataError("Invalid path for byron address!")
 
+    address = derive_byron_address(keychain, path, protocol_magic)
+    return address
 
-def validate_full_path(path: list) -> bool:
+
+def _derive_shelley_address(
+    keychain: seed.Keychain, parameters: CardanoAddressParametersType, network_id: int,
+) -> bytes:
+    if not is_shelley_path(parameters.spending_key_path):
+        raise wire.DataError("Invalid path for shelley address!")
+
+    address_header = _create_address_header(parameters.address_type, network_id)
+
+    if parameters.address_type == CardanoAddressType.BASE:
+        address = _derive_base_address(
```

Comment:
please pass the `network_id` into the `_derive_*_address` functions, AND set the appropriate `address_type` value inside (derived `base` address must necessarily have type `CardanoAddressType.BASE` so it should be constant inside that function)

Reply:
Done [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/76bc4dcc3a298bc190d0cd4263fae7ed1d998d73).